tests/src/test/java/org/eclipse/hono/tests/client/ClientTestHelper.java

```diff
@@ -0,0 +1,130 @@
+/**
+ * Copyright (c) 2018 Bosch Software Innovations GmbH and others.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ */
+package org.eclipse.hono.tests.client;
+
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Consumer;
+
+import org.apache.qpid.proton.message.Message;
+import org.eclipse.hono.client.MessageConsumer;
+import org.eclipse.hono.client.MessageSender;
+import org.eclipse.hono.tests.IntegrationTestSupport;
+import org.eclipse.hono.util.MessageHelper;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.vertx.core.Future;
+import io.vertx.core.Handler;
+import io.vertx.ext.unit.Async;
+import io.vertx.ext.unit.TestContext;
+
+/**
+ * Base class for integration tests for Hono's AMQP 1.0 services.
+ */
+public abstract class ClientTestHelper {
+
+    /**
+     * A logger to be used by subclasses.
+     */
+    protected final Logger LOGGER = LoggerFactory.getLogger(getClass());
+
+    private static final long   DEFAULT_TEST_TIMEOUT = 15000; // ms
+
+    protected MessageSender sender;
+    protected MessageConsumer consumer;
+
```

Comment:
these fields are never used locally, are they? Which method is responsible for initializing them and closing the links later?

Reply:
> these fields are never used locally, are they?

Yes, they are used by subclasses.

> Which method is responsible for initializing them

They are initialized either before a test executes (`@Before` in the `ClientTestBase`) or during test execution (that's what the `setupProtocolAdapter` does). the creation and initialization is left to subclasses

> and closing the links later?

Every subclass runs its own `@After` for closing the links. so the closing of the links is left to subclasses

Reply:
How is the implementor of a subclass supposed to know about all of this? The *createConsumer* and *createProducer* methods seem to indicate that a subclass is only responsible for creating these objects. However, the JavaDocs of those methods do not indicate that the respective result should be used to initialize the fields. And neither does it instruct implementors to close those objects once they are done. Adding these instructions to the JavaDoc, on the other hand, will still not make for a strong contract, because everything is dependent on the implementor's ability to follow the rules. IMHO neither the fields not the methods should be declared on ClientTestHelper. Instead, the *doUploadMessages* method should probably accept a sender and consumer as params ...

Reply:
yep, i see now. actually, the reason i placed them in the parent is to avoid declaring the same fields in both subclass (but i didnt see the bigger picture :+1: ). thanks..will update the PR